Separate system tests out of E2E

mkolesnik opened the pull request:
This PR separates the system tests out of E2E, making E2E super simple and system tests more focused.

As part of this work the E2E tests will be the default ones (as run by Shipyard).
The system tests have been improved a lot as part of this effort.

Please see individual commits for more details.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

mkolesnik committed:
Run E2E even on draft

While it made (some) sense on operator, there's no need to have this
restriction on `subctl` repo as it's quite useful to run the default E2E
on drafts.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Move `e2e` to `system`

We have E2E already, these are much more "system" tests

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Remove E2E testing from system tests

The system tests should focus on testing the functionality of `subctl`
so removing all E2E testing which is already covered by the `E2E` type
of tests.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Fixup system tests and add to GHA

Fixup system tests and add them to GHA

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Simplify system tests

Simplify by adding a function to wrap calls which also prints helpful
GHA sections.

Also most if the commands support `$KUBECONFIG` so use that + contexts
when possible instead of explicitly specifying

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Fold `test-subctl` into system tests

Most are already covered by the system tests, the ones that weren't were
folded into the system tests.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

mkolesnik committed:
Use short FW timeout on system tests

Since we're testing using KIND, latency isn't expected (network is local
to the machine). Hence, we can use short timeouts to have the tests run
much faster (a few good minutes reduction overall).

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

tpantelis commented on .github/workflows/system.yml:
```diff
@@ -1,22 +1,27 @@
-  test-subctl:
-    name: subctl kubeconfig commands
+  system:
+    name: Functionality
```
```suggestion
    name: Execute subctl commands
```

mkolesnik committed:
Use smaller KIND clusters

As we're basically running standard E2E without any special
requirements, we can use smaller clusters and save some time and
resources when deploying with KIND.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

skitt commented on Makefile:
```diff
@@ -75,8 +76,8 @@ export PATH := $(CURDIR)/cmd/bin:$(PATH)
 
-e2e: deploy
-	scripts/kind-e2e/e2e.sh $(E2E_ARGS)
+system: deploy
```
`system` on its own doesn’t convey the meaning of the target; I think `test`, or `system-test`, would be clearer. Since the goal is to test `subctl` (as illustrated by subsequent commits), `test` feels appropriate.

skitt replied:
I don’t see the fix…

skitt commented on Makefile:
```diff
@@ -75,8 +76,8 @@ export PATH := $(CURDIR)/cmd/bin:$(PATH)
 
-e2e: deploy
-	scripts/kind-e2e/e2e.sh $(E2E_ARGS)
+system:
```
This still says `system`.

mkolesnik committed:
Rename `system` target to `system-test`

Since `system` on its own doesn't convey the meaning of the target,
`system-test` makes it clear what this is doing.

Signed-off-by: Mike Kolesnik <mkolesni@redhat.com>

dfarrell07 commented on scripts/kind-e2e/lib_operator_verify_subm.sh:
```diff
@@ -1,491 +0,0 @@
-  kubectl get pod $subm_gateway_pod_name --namespace=$subm_ns -o json | tee $json_file
-
-  validate_pod_container_equals 'image' "${subm_gateway_image_repo}/submariner-gateway:local"
-  validate_pod_container_has 'securityContext.capabilities.add' 'net_admin'
```
Am I right in understanding that the new tests don't actually assert details at this level? They just check for the presence of K8s objects?

mkolesnik replied:
Yes as this knowledge is internal to operator and tested there, here its not necessary nor helpful to have these white box tests as the operator takes care of these details.

dfarrell07 replied:
Ah so we're keeping these tests in https://github.com/submariner-io/submariner-operator/pull/2104 but removing them here, I see and that makes sense.
